tracking/src/Tracker.js

```diff
@@ -0,0 +1,126 @@
+const pjson = require('../../package.json')
+const BASE_URL = "https://analytics.data.visenze.com/v3";
+const BASE_URL_CN = "https://analytics.visenze.com.cn/v3";
+const SDK = "javascript sdk"; 
+const SDK_VERSION = pjson.version; 
+const TIMEOUT = 15000;
+
+
+/**
+* Adds a list of query parameters
+* @param  {params}  params object
+* @return {URI}     returns self for fluent chaining
+*/
+URI.prototype.addQueryParams = function (params) {
+    for (const property in params) {
+      if (params.hasOwnProperty(property)) {
+        const param = params[property];
+        // do stuff
+        if (Array.isArray(param)) {
+          for (let i = 0; i < param.length; i += 1) {
+            this.addQueryParam(property, param[i]);
+          }
+        } else {
+          this.addQueryParam(property, param);
+        }
+      }
+    }
+    return this;
+};
+
+const timeout = (ms, promise) => {
+    return new Promise((resolve, reject) => {
+        setTimeout(() => {
+          reject(`Timed out in ${ms} ms.`);
+        }, ms);
+        return promise.then(resolve, reject);
+    });
+}
+
+
+
+const sendRequest = (fetchObj, path, callback, failure) => {
+    const start = new Date().getTime(); 
+    return timeout(TIMEOUT, fetchObj)
+        .then(response => {
+            const status = response.status; 
+            if(status === 200) {
+                return "success"; 
+            } else {
+                const errResp = response.json(); 
+                return errResp; 
+            }
+        })
+        .then(res => {
+            if (res === "success") {
+                if(callback) {
+                    callback("send success")
+                }
+            } else {
+                if (failure) {
+                    failure(res);
+                }
+            }
+        })
+        .catch(ex => {
+            console.error(`Failed to process ${path}`, ex);
+            if (failure) {
+              failure(ex);
+            }                
+        })
+}
+
+const sendGetRequest = (path, params, callback, failure) => {
+    const url = new URI(path)
+        .addQueryParams(params)
+        .toString(); 
+    console.log("url: ", url);    
+    const fetchObj = fetch(url, {
+        method: 'GET'
+    });
+
+    return sendRequest(fetchObj, path, callback, failure); 
+}
+
+
+class Tracker {
+    constructor(dataCollection, sessionManager, code, country) {
+        this.code = code; 
+        this.dataCollection = dataCollection; 
+        this.sessionManager = sessionManager; 
+        this.uid = sessionManager.getUserID(); 
+        this.country = country; 
+        if(country === "CN") {
+            this.baseUrl = BASE_URL_CN; 
+        } else {
+            this.baseUrl = BASE_URL; 
+        }
+    }
+
+    sendEvent(eventName, dataObj, callback) {
+        const path = `${this.baseUrl}/__va.gif`; 
+        let defaultsParams = this.dataCollection.toJson(); 
+        let params = Object.assign(defaultsParams, dataObj); 
```

Comment:
we should not be allowing the end user to override those auto-capture params (e.g os,platform,etc..) right?

so it should be `Object.assign(dataObj, defaultsParams);`?

Reply:
i think the auto-captured paramters are not accurate. i copy the logic from here: 
https://github.com/fingerprintjs/fingerprintjs2/blob/master/fingerprint2.js 

like UID, i think if user have provide it as parameter, then need to take user input parameter as the final one. 

@thehung111  please confirm the logic.

Reply:
but you already overriding some of the user input parameters below
Example: the `uid` will be overridden by the session manager
```
       params.code = this.code; 
        params.uid = this.sessionManager.getUserID(); 
        params.sid = this.sessionManager.getSessionId(); 
        params.sdk = SDK; 
        params.v = SDK_VERSION; 
```

Reply:
emm yes. you are right,  i should not override. user input.  uid should not override. in case user just want to change uid in every event. 
but sessionId. i need to override. cause it is generate with out logic. sdk and v, i think i should also override. cause it is library itself related. if not user can just mess up with sdk and version easily i think it is not what we want .

Reply:
which parameters are not accurate?

Reply:
data colleted from user-agent is not accurate. cause i see firngerprint.js they have additional logic to judge whether it can be trusted. anyway since backend already handled data parsing from user agent. i have removed the data collection from user agent. just to avoid the data been overwritten

Reply:
ok